src/Sparrow/Utils/ZstdStream.cs

```diff
@@ -181,60 +201,104 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
-                await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
+                    await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
+
+                }
             }
+            finally
+            {
+                _semaphoreSlim.Release();
+            }
+
         }
 
         public override void Flush()
         {
-            if (_disposeOnce.Disposed)
-                throw new ObjectDisposedException("Object was already disposed!");
+            WaitOnSemaphoreOrThrowIfDisposed();
 
-            while (true)
+            try
+            {
+                while (true)
+                {
+                    var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
+                    if (outputBytes == 0)
+                        break;
+                    _inner.Write(_tempBuffer, 0, outputBytes);
+                }
+                _inner?.Flush();
+            }
+            finally
             {
-                var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
-                if (outputBytes == 0)
-                    break;
-                _inner.Write(_tempBuffer, 0, outputBytes);
+                _semaphoreSlim.Release();
             }
-            _inner?.Flush();
         }
 
         public override async Task FlushAsync(CancellationToken cancellationToken)
         {
-            if (_disposeOnce.Disposed)
-                throw new ObjectDisposedException("Object was already disposed!");
+            await WaitOnSemaphoreOrThrowIfDisposedAsync().ConfigureAwait(false);
 
-            while (true)
+            try
             {
-                var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
-                if (outputBytes == 0)
-                    break;
+                while (true)
+                {
+                    var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
+                    if (outputBytes == 0)
+                        break;
 
-                await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
+                    await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
+                }
+                await _inner.FlushAsync(cancellationToken).ConfigureAwait(false);
             }
-            await _inner.FlushAsync(cancellationToken).ConfigureAwait(false);
+            finally
+            {
+                _semaphoreSlim.Release();
+            }
+        }
+
+        private void WaitOnSemaphoreOrThrowIfDisposed()
+        {
+            if (_disposeOnce.Disposed)
+                throw new ObjectDisposedException("Object was already disposed!");
+
+            _semaphoreSlim.Wait();
+        }
```

Comment:
And now you are racy between checking the dispose and the actual usage, no?

Reply:
Would it be better to use this: `https://github.com/ravendb/ravendb/blob/v5.3/src/Sparrow.Server/Utils/DisposeLock.cs` ?

Reply:
done